src/tiered/tiered_handle.c

```diff
@@ -8,63 +8,488 @@
 
 #include "wt_internal.h"
 
+/*
+ * __tiered_init_tiers --
+ *     Given a tiered table 'tiers' configuration set up the dhandle array.
+ */
+static int
+__tiered_init_tiers(WT_SESSION_IMPL *session, WT_TIERED *tiered, WT_CONFIG_ITEM *tierconf)
+{
+    WT_CONFIG cparser;
+    WT_CONFIG_ITEM ckey, cval;
+    WT_DECL_ITEM(tmp);
+    WT_DECL_RET;
+    uint32_t type;
+
+    WT_RET(__wt_scr_alloc(session, 0, &tmp));
+    __wt_config_subinit(session, &cparser, tierconf);
+    while ((ret = __wt_config_next(&cparser, &ckey, &cval)) == 0) {
+        /* Set up the tiers array based on the metadata. */
+        WT_ERR(__wt_buf_fmt(session, tmp, "%.*s", (int)ckey.len, ckey.str));
+        __wt_verbose(
+          session, WT_VERB_TIERED, "INIT_TIERS: tiered URI dhandle %s", (char *)tmp->data);
+        WT_ERR(__wt_session_get_dhandle(session, (const char *)tmp->data, NULL, NULL, 0));
+        if (session->dhandle->type == WT_DHANDLE_TYPE_BTREE)
+            tiered->tiers[WT_TIERED_INDEX_WRITEABLE] = session->dhandle;
+        else if (session->dhandle->type == WT_DHANDLE_TYPE_TIERED)
+            tiered->tiers[WT_TIERED_INDEX_READONLY] = session->dhandle;
+        else {
+            type = (uint32_t)session->dhandle->type;
+            WT_TRET(__wt_session_release_dhandle(session));
+            WT_ERR_MSG(session, EINVAL, "Unknown or unsupported tiered dhandle type %d", (int)type);
+        }
+        (void)__wt_atomic_addi32(&session->dhandle->session_inuse, 1);
+        __wt_verbose(session, WT_VERB_TIERED, "INIT_TIERS: DHANDLE %s inuse %d",
+          session->dhandle->name, session->dhandle->session_inuse);
+        WT_ERR(__wt_session_release_dhandle(session));
+    }
+    WT_ERR_NOTFOUND_OK(ret, false);
+err:
+    __wt_scr_free(session, &tmp);
+    return (ret);
+}
+
+/*
+ * __tiered_create_local --
+ *     Create a new local name for a tiered table.
+ */
+static int
+__tiered_create_local(WT_SESSION_IMPL *session, WT_TIERED *tiered)
+{
+
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_ITEM(tmp);
+    WT_DECL_RET;
+    const char *cfg[4] = {NULL, NULL, NULL, NULL};
+    const char *config, *name;
+
+    conn = S2C(session);
+    WT_RET(__wt_scr_alloc(session, 0, &tmp));
+
+    /* This is the only place that touches this value. No need for atomic operation. */
```

Comment:
I would add here or in function comment that this must be called single threaded.